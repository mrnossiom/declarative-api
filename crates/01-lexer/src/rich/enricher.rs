use crate::{
	poor,
	rich::{AttrStyle, Delimiter, LiteralKind, OpKind, Token, TokenKind},
};
use session::{BytePos, ParseSession, SourceFile, Span, Symbol};
use tracing::instrument;

/// Transforms [`poor::Token`]s that are only relevant when reading the source file at the
/// same time into [`rich::Token`](crate::rich::Token)s that are self-explanatory. The latter doesn't include
/// tokens that don't add information to the generator such as whitespace or comments.
pub struct Enricher<'a> {
	session: &'a ParseSession,
	source: &'a str,
	cursor: poor::Cursor<'a>,
	start_pos: BytePos,
	pos: BytePos,
}

impl<'a> Enricher<'a> {
	/// Creates a new [`Enricher`] for the given source, creating in the way
	/// the underlying [`poor::Cursor`] to get tokens to enrich.
	#[must_use]
	pub fn from_source(session: &'a ParseSession, source: &'a SourceFile) -> Self {
		Self {
			session,
			source: &source.source,
			cursor: poor::Cursor::from_source(&source.source),
			start_pos: source.start_pos,
			pos: source.start_pos,
		}
	}

	/// Fetches the next token from the underlying poor lexer and add information
	///
	/// # Panics
	/// TODO: invalid characters or ident are not implemented yet
	#[instrument(level = "DEBUG", skip(self))]
	pub fn next_token(&mut self) -> (Token, bool) {
		let mut has_whitespace_before = false;

		loop {
			let token = self.cursor.advance_token();
			let start = self.pos;
			self.pos += BytePos(token.length);

			let kind = match token.kind {
				poor::TokenKind::Ident => {
					let ident = self.str_from(start);
					TokenKind::Ident(Symbol::intern(ident))
				}
				poor::TokenKind::LineComment(style) => {
					// Skip non-doc comments
					let Some(style) = style else {
						has_whitespace_before = true;
						continue;
					};

					let len = match style {
						poor::DocStyle::Inner => 3, // `##!`
						poor::DocStyle::Outer => 2, // `##`
					};

					let content = self.str_from(start + BytePos(len));

					let (style, sym) = Self::cook_doc_line_comment(content, style);

					TokenKind::DocComment(style, sym)
				}
				poor::TokenKind::Literal(kind) => {
					let (kind, sym) = self.cook_literal(start, self.pos, kind);

					TokenKind::Literal(kind, sym)
				}

				poor::TokenKind::Whitespace => {
					has_whitespace_before = true;
					continue;
				}

				poor::TokenKind::InvalidIdent | poor::TokenKind::Unknown => todo!(),

				poor::TokenKind::Semi => TokenKind::Semi,
				poor::TokenKind::Comma => TokenKind::Comma,
				poor::TokenKind::Dot => TokenKind::Dot,

				poor::TokenKind::OpenParenthesis => TokenKind::OpenDelim(Delimiter::Parenthesis),
				poor::TokenKind::CloseParenthesis => TokenKind::CloseDelim(Delimiter::Parenthesis),
				poor::TokenKind::OpenBrace => TokenKind::OpenDelim(Delimiter::Brace),
				poor::TokenKind::CloseBrace => TokenKind::CloseDelim(Delimiter::Brace),
				poor::TokenKind::OpenBracket => TokenKind::OpenDelim(Delimiter::Bracket),
				poor::TokenKind::CloseBracket => TokenKind::CloseDelim(Delimiter::Bracket),

				poor::TokenKind::At => TokenKind::At,
				poor::TokenKind::Pound => TokenKind::Pound,
				poor::TokenKind::Tilde => TokenKind::Tilde,
				poor::TokenKind::Question => TokenKind::Question,
				poor::TokenKind::Colon => TokenKind::Colon,
				poor::TokenKind::Dollar => TokenKind::Dollar,
				poor::TokenKind::Eq => TokenKind::Eq,
				poor::TokenKind::Bang => TokenKind::Bang,

				poor::TokenKind::Lt => TokenKind::Op(OpKind::Lt),
				poor::TokenKind::Gt => TokenKind::Op(OpKind::Gt),
				poor::TokenKind::Minus => TokenKind::Op(OpKind::Minus),
				poor::TokenKind::And => TokenKind::Op(OpKind::And),
				poor::TokenKind::Or => TokenKind::Op(OpKind::Or),
				poor::TokenKind::Plus => TokenKind::Op(OpKind::Plus),
				poor::TokenKind::Star => TokenKind::Op(OpKind::Star),
				poor::TokenKind::Slash => TokenKind::Op(OpKind::Slash),
				poor::TokenKind::Caret => TokenKind::Op(OpKind::Caret),
				poor::TokenKind::Percent => TokenKind::Op(OpKind::Percent),

				poor::TokenKind::Eof => TokenKind::Eof,
			};

			let span = Span::from_bounds(start, self.pos);
			return (Token::new(kind, span), has_whitespace_before);
		}
	}

	#[instrument(level = "DEBUG")]
	fn cook_doc_line_comment(content: &str, style: poor::DocStyle) -> (AttrStyle, Symbol) {
		let style = match style {
			poor::DocStyle::Inner => AttrStyle::Inner,
			poor::DocStyle::Outer => AttrStyle::OuterOrInline,
		};

		(style, Symbol::intern(content))
	}

	#[instrument(level = "DEBUG", skip(self))]
	fn cook_literal(
		&mut self,
		start: BytePos,
		end: BytePos,
		kind: poor::LiteralKind,
	) -> (LiteralKind, Symbol) {
		match kind {
			poor::LiteralKind::Str { terminated } => {
				if !terminated {
					// error somehow
				}

				let content = self.str_from_to(start + BytePos(1), end - BytePos(1));
				(LiteralKind::Str, Symbol::intern(content))
			}
			poor::LiteralKind::Number => {
				let content = self.str_from(start);
				(LiteralKind::Number, Symbol::intern(content))
			}
		}
	}

	#[inline]
	fn src_index(&self, pos: BytePos) -> usize {
		(pos - self.start_pos).0 as usize
	}

	/// Slice of the source text from `start` up to but excluding `self.pos`,
	/// meaning the slice does not include the character `self.ch`.
	#[instrument(level = "DEBUG", skip(self))]
	fn str_from(&self, start: BytePos) -> &'a str {
		self.str_from_to(start, self.pos)
	}

	/// Slice of the source text spanning from `start` up to but excluding `end`.
	#[instrument(level = "DEBUG", skip(self))]
	fn str_from_to(&self, start: BytePos, end: BytePos) -> &'a str {
		&self.source[self.src_index(start)..self.src_index(end)]
	}
}

pub struct Iter<'a>(Enricher<'a>);

impl<'a> Iterator for Iter<'a> {
	type Item = Token;

	fn next(&mut self) -> Option<Self::Item> {
		let (token, _) = self.0.next_token();

		if token.kind.is_eof() {
			None
		} else {
			Some(token)
		}
	}
}

impl<'a> IntoIterator for Enricher<'a> {
	type Item = Token;
	type IntoIter = Iter<'a>;

	fn into_iter(self) -> Self::IntoIter {
		Iter(self)
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::tests::{ATTR, EXAMPLE, URLS};
	use session::symbols::attrs;

	macro_rules! sym {
		($lit:literal) => {
			Symbol::intern($lit)
		};
	}

	macro_rules! tokens {
		($expr:ident, $(($ty:expr, [$lo:literal, $hi:literal])),+) => {
			let psession = ParseSession::default();
			let source = psession.source_map.add_source($expr.into());
			let mut tokens = Enricher::from_source(&psession, &source).into_iter();

			$(
				assert_eq!(
					tokens.next(),
					Some(Token::new($ty, Span::from_bounds(BytePos($lo), BytePos($hi))))
				);
			)+
			assert_eq!(tokens.next(), None);
		};
	}

	#[test]
	fn can_enrich_example_file() {
		let parse_sess = ParseSession::default();
		let source = parse_sess.source_map.add_source(EXAMPLE.into());

		let rich_tokens = Enricher::from_source(&parse_sess, &source)
			.into_iter()
			.collect::<Vec<_>>();

		println!("{rich_tokens:#?}");
	}

	#[test]
	fn parse_attr() {
		use crate::rich::TokenKind::*;
		use session::Span;

		tokens!(
			ATTR,
			(At, [0, 1]),
			(Ident(attrs::Format), [1, 7]),
			(Colon, [7, 8]),
			(Ident(Symbol::intern("date")), [9, 13])
		);
	}

	#[test]
	fn parse_array_like() {
		use crate::rich::{Delimiter::*, LiteralKind::*, TokenKind::*};
		use session::Span;

		tokens![
			URLS,
			(Ident(sym!("urls")), [0, 4]),
			(OpenDelim(Bracket), [5, 6]),
			(
				Literal(Str, sym!("https://paradigm.lighton.ai/api/v1")),
				[8, 44]
			),
			(
				Literal(Str, sym!("https://paradigm-preprod.lighton.ai/api/v1")),
				[46, 90]
			),
			(
				Literal(Str, sym!("https://paradigm-dev.lighton.ai/api/v1")),
				[92, 132]
			),
			(CloseDelim(Bracket), [133, 134])
		];
	}
}
