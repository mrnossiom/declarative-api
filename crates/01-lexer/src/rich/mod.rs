mod enricher;
mod errors;
mod token;

pub use enricher::Enricher;
pub use token::*;
