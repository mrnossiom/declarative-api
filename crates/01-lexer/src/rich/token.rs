use session::{Ident, Span, Symbol};
use std::fmt;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Token {
	pub kind: TokenKind,
	pub span: Span,
}

impl Token {
	pub const DUMMY: Self = Self {
		kind: TokenKind::Eof,
		span: Span::DUMMY,
	};

	pub(crate) const fn new(kind: TokenKind, span: Span) -> Self {
		Self { kind, span }
	}

	#[must_use]
	pub const fn ident(&self) -> Option<Ident> {
		match &self.kind {
			TokenKind::Ident(sym) => Some(Ident::new(*sym, self.span)),
			_ => None,
		}
	}

	#[must_use]
	pub fn is_keyword(&self, kw: Symbol) -> bool {
		self.ident().map_or(false, |id| id.symbol == kw)
	}
}

impl fmt::Display for Token {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		write!(f, "{} ({})", self.kind, self.span)
	}
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TokenKind {
	// IDEA: maybe multiline
	DocComment(AttrStyle, Symbol),

	/// An identifier
	Ident(Symbol),

	/// Like the above, but containing invalid unicode codepoints.
	InvalidIdent,

	/// See [LiteralKind] for more details.
	Literal(LiteralKind, Symbol),

	// One-char tokens:
	/// ";"
	Semi,
	/// ","
	Comma,
	/// "."
	Dot,

	/// An opening delimiter (e.g., `{`).
	OpenDelim(Delimiter),
	/// A closing delimiter (e.g., `}`).
	CloseDelim(Delimiter),

	/// "@"
	At,
	/// "#"
	Pound,
	/// "~"
	Tilde,
	/// "?"
	Question,
	/// ":"
	Colon,
	/// "$"
	Dollar,
	/// "="
	Eq,
	/// "!"
	Bang,

	Op(OpKind),

	/// Unknown token, not expected by the lexer, e.g. "№"
	Unknown,

	/// End of input.
	Eof,
}

impl TokenKind {
	#[must_use]
	pub const fn is_eof(&self) -> bool {
		matches!(self, Self::Eof)
	}
}

impl fmt::Display for TokenKind {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		match self {
			Self::DocComment(style, sym) => write!(
				f,
				r#"{} doc comment ("{sym}")"#,
				match style {
					AttrStyle::OuterOrInline => "outer",
					AttrStyle::Inner => "inner",
				}
			),

			Self::Ident(sym) => write!(f, r#"ident "{sym}""#),
			Self::InvalidIdent => write!(f, "invalid ident"),
			Self::Literal(LiteralKind::Bool, _sym) => {
				unreachable!("bool literal doesn't exist in lexer")
			}
			Self::Literal(LiteralKind::Number, sym) => write!(f, "lit {sym}"),
			Self::Literal(LiteralKind::Str, sym) => write!(f, r#"lit "{sym}""#),

			Self::Semi => write!(f, ";"),
			Self::Comma => write!(f, ","),
			Self::Dot => write!(f, "."),
			Self::OpenDelim(Delimiter::Parenthesis) => write!(f, "("),
			Self::CloseDelim(Delimiter::Parenthesis) => write!(f, ")"),
			Self::OpenDelim(Delimiter::Brace) => write!(f, "{{"),
			Self::CloseDelim(Delimiter::Brace) => write!(f, "}}"),
			Self::OpenDelim(Delimiter::Bracket) => write!(f, "["),
			Self::CloseDelim(Delimiter::Bracket) => write!(f, "]"),
			Self::OpenDelim(Delimiter::Invisible) => write!(f, "Ø..."),
			Self::CloseDelim(Delimiter::Invisible) => write!(f, "...Ø"),

			Self::At => write!(f, "@"),
			Self::Pound => write!(f, "#"),
			Self::Tilde => write!(f, "~"),
			Self::Question => write!(f, "?"),
			Self::Colon => write!(f, ":"),
			Self::Dollar => write!(f, "$"),
			Self::Eq => write!(f, "="),
			Self::Bang => write!(f, "!"),

			Self::Op(op) => write!(f, "{op}"),

			Self::Unknown => write!(f, "unknown"),

			Self::Eof => write!(f, "<EOF>"),
		}
	}
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LiteralKind {
	/// AST only, represent  `true` or `false`
	Bool,

	/// ""abc"", ""abc"
	Str,
	/// "12_u8", "0o100", "0b120i99", "1f32".
	Number,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum Delimiter {
	/// `( ... )`
	Parenthesis,
	/// `{ ... }`
	Brace,
	/// `[ ... ]`
	Bracket,
	/// `Ø ... Ø`
	/// An invisible delimiter, that may, for example, appear around tokens coming from a
	/// "macro variable" `$var`. It is important to preserve operator priorities in cases like
	/// `$var * 3` where `$var` is `1 + 2`.
	/// Invisible delimiters might not survive roundtrip of a token stream through a string.
	Invisible,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
pub enum OpKind {
	/// "<"
	Lt,
	/// ">"
	Gt,
	/// "-"
	Minus,
	/// "&"
	And,
	/// "|"
	Or,
	/// "+"
	Plus,
	/// "*"
	Star,
	/// "/"
	Slash,
	/// "^"
	Caret,
	/// "%"
	Percent,
}

impl fmt::Display for OpKind {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		match self {
			Self::Lt => write!(f, ">"),
			Self::Gt => write!(f, "<"),
			Self::Minus => write!(f, "-"),
			Self::And => write!(f, "&"),
			Self::Or => write!(f, "|"),
			Self::Plus => write!(f, "+"),
			Self::Star => write!(f, "*"),
			Self::Slash => write!(f, "/"),
			Self::Caret => write!(f, "^"),
			Self::Percent => write!(f, "%"),
		}
	}
}

/// An attribute of these forms
///
/// `@name(<tokens>)`
/// `@name{<tokens>}`
/// `@name[<tokens>]`
/// `@name: <token>`
///
/// or an outer doc comment
///
/// `## doc comment`
///
/// with an optional `!` after the `@` or the `##` to signify inner.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AttrStyle {
	/// Without a `!` bang
	OuterOrInline,

	/// With a `!` bang
	Inner,
}
