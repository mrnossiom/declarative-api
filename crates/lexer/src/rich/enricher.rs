use crate::{
	poor,
	rich::{errors, Delimiter, DocStyle, LiteralKind, OpKind, Token, TokenKind},
};
use dapic_session::{BytePos, ParseSession, SourceFile, Span, Symbol};
use tracing::instrument;

/// Transforms [`poor::Token`]s that are only relevant when reading the source file at the
/// same time into [`rich::Token`](crate::rich::Token)s that are self-explanatory. The latter doesn't include
/// tokens that don't add information to the generator such as whitespace or comments.
pub struct Enricher<'gctx> {
	session: &'gctx ParseSession<'gctx>,
	source: &'gctx str,
	cursor: poor::Cursor<'gctx>,
	start_pos: BytePos,
	pos: BytePos,
}

impl<'a> Enricher<'a> {
	/// Creates a new [`Enricher`] for the given source, creating in the way
	/// the underlying [`poor::Cursor`] to get tokens to enrich.
	#[must_use]
	pub fn from_source(session: &'a ParseSession<'a>, source: &'a SourceFile) -> Self {
		Self {
			session,
			source: &source.source,
			cursor: poor::Cursor::from_source(&source.source),
			start_pos: source.offset,
			pos: source.offset,
		}
	}

	/// Fetches the next token from the underlying poor lexer and add information
	#[instrument(level = "DEBUG", skip(self))]
	pub fn next_token(&mut self) -> (Token, bool) {
		let mut has_whitespace_before = false;

		loop {
			let token = self.cursor.next_token();
			let start = self.pos;
			self.pos = self.pos + BytePos(token.length);

			let span = Span::from_bounds(start, self.pos);

			let kind = match token.kind {
				poor::TokenKind::Ident => {
					let ident = self.str_from(start);
					TokenKind::Ident(Symbol::intern(ident))
				}
				poor::TokenKind::LineComment(style) => {
					// Skip non-doc comments
					let Some(style) = style else {
						has_whitespace_before = true;
						continue;
					};

					let len = match style {
						DocStyle::Inner => 3, // `##!`
						DocStyle::Outer => 2, // `##`
					};

					let content = self.str_from(start + BytePos(len));

					let (style, sym) = Self::cook_doc_line_comment(content, style);

					TokenKind::DocComment(style, sym)
				}
				poor::TokenKind::Literal(kind) => {
					let (kind, sym) = self.cook_literal(start, self.pos, kind);

					TokenKind::Literal(kind, sym)
				}

				poor::TokenKind::Whitespace => {
					has_whitespace_before = true;
					continue;
				}

				poor::TokenKind::InvalidIdent => {
					self.session.diag.emit(errors::InvalidIdent { span });

					// TODO: enhance recovery by emitting a wrong identifier or helping the user to change it

					has_whitespace_before = true;
					continue;
				}

				poor::TokenKind::Unknown => {
					self.session.diag.emit(errors::Unknown { span });

					has_whitespace_before = true;
					continue;
				}

				poor::TokenKind::Semi => TokenKind::Semi,
				poor::TokenKind::Comma => TokenKind::Comma,
				poor::TokenKind::Dot => TokenKind::Dot,

				poor::TokenKind::OpenParenthesis => TokenKind::OpenDelim(Delimiter::Parenthesis),
				poor::TokenKind::CloseParenthesis => TokenKind::CloseDelim(Delimiter::Parenthesis),
				poor::TokenKind::OpenBrace => TokenKind::OpenDelim(Delimiter::Brace),
				poor::TokenKind::CloseBrace => TokenKind::CloseDelim(Delimiter::Brace),
				poor::TokenKind::OpenBracket => TokenKind::OpenDelim(Delimiter::Bracket),
				poor::TokenKind::CloseBracket => TokenKind::CloseDelim(Delimiter::Bracket),

				poor::TokenKind::At => TokenKind::At,
				poor::TokenKind::Pound => TokenKind::Pound,
				poor::TokenKind::Tilde => TokenKind::Tilde,
				poor::TokenKind::Question => TokenKind::Question,
				poor::TokenKind::Colon => TokenKind::Colon,
				poor::TokenKind::Dollar => TokenKind::Dollar,
				poor::TokenKind::Eq => TokenKind::Eq,
				poor::TokenKind::Bang => TokenKind::Bang,

				poor::TokenKind::Lt => TokenKind::Op(OpKind::Lt),
				poor::TokenKind::Gt => TokenKind::Op(OpKind::Gt),
				poor::TokenKind::Minus => TokenKind::Op(OpKind::Minus),
				poor::TokenKind::And => TokenKind::Op(OpKind::And),
				poor::TokenKind::Or => TokenKind::Op(OpKind::Or),
				poor::TokenKind::Plus => TokenKind::Op(OpKind::Plus),
				poor::TokenKind::Star => TokenKind::Op(OpKind::Star),
				poor::TokenKind::Slash => TokenKind::Op(OpKind::Slash),
				poor::TokenKind::Caret => TokenKind::Op(OpKind::Caret),
				poor::TokenKind::Percent => TokenKind::Op(OpKind::Percent),

				poor::TokenKind::Eof => TokenKind::Eof,
			};

			return (Token::new(kind, span), has_whitespace_before);
		}
	}

	#[instrument(level = "DEBUG")]
	fn cook_doc_line_comment(content: &str, style: DocStyle) -> (DocStyle, Symbol) {
		(style, Symbol::intern(content))
	}

	#[instrument(level = "DEBUG", skip(self))]
	fn cook_literal(
		&mut self,
		start: BytePos,
		end: BytePos,
		kind: poor::LiteralKind,
	) -> (LiteralKind, Symbol) {
		match kind {
			poor::LiteralKind::Str { terminated } => {
				if !terminated {
					// error somehow
				}

				let content = self.str_from_to(start + BytePos(1), end - BytePos(1));
				(LiteralKind::Str, Symbol::intern(content))
			}
			poor::LiteralKind::Number => {
				let content = self.str_from(start);
				(LiteralKind::Number, Symbol::intern(content))
			}
		}
	}

	#[inline]
	fn src_index(&self, pos: BytePos) -> usize {
		(pos - self.start_pos).0 as usize
	}

	/// Slice of the source text from `start` up to but excluding `self.pos`,
	/// meaning the slice does not include the character `self.ch`.
	#[instrument(level = "DEBUG", skip(self))]
	fn str_from(&self, start: BytePos) -> &'a str {
		self.str_from_to(start, self.pos)
	}

	/// Slice of the source text spanning from `start` up to but excluding `end`.
	#[instrument(level = "DEBUG", skip(self))]
	fn str_from_to(&self, start: BytePos, end: BytePos) -> &'a str {
		&self.source[self.src_index(start)..self.src_index(end)]
	}
}

pub struct Iter<'a>(Enricher<'a>);

impl<'a> Iterator for Iter<'a> {
	type Item = Token;

	fn next(&mut self) -> Option<Self::Item> {
		let (token, _) = self.0.next_token();

		if token.kind.is_eof() {
			None
		} else {
			Some(token)
		}
	}
}

impl<'a> IntoIterator for Enricher<'a> {
	type Item = Token;
	type IntoIter = Iter<'a>;

	fn into_iter(self) -> Self::IntoIter {
		Iter(self)
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::tests::{ATTR, EXAMPLE, URLS};
	use dapic_session::{sym, symbols::attrs, Session};

	macro_rules! tokens {
		($expr:ident, $(($ty:expr, [$lo:literal, $hi:literal])),+) => {
			let sess = Session::default();
			let source = sess.source_map.load_anon($expr.into());
			let psess = sess.parse_sess();
			let mut tokens = Enricher::from_source(&psess, &source).into_iter();

			$(
				assert_eq!(
					tokens.next(),
					Some(Token::new($ty, dapic_session::sp!($lo, $hi)))
				);
			)+
			assert_eq!(tokens.next(), None);
		};
	}

	#[test]
	fn can_enrich_example_file() {
		let sess = Session::default();
		let source = sess.source_map.load_anon(EXAMPLE.into());

		let rich_tokens = Enricher::from_source(
			&{
				let this = &sess;
				ParseSession {
					diag: &this.diagnostics,
				}
			},
			&source,
		)
		.into_iter()
		.collect::<Vec<_>>();

		println!("{rich_tokens:#?}");
	}

	#[test]
	fn parse_attr() {
		use crate::rich::TokenKind::*;

		tokens!(
			ATTR,
			(At, [0, 1]),
			(Ident(attrs::format), [1, 7]),
			(Colon, [7, 8]),
			(Ident(Symbol::intern("date")), [9, 13])
		);
	}

	#[test]
	fn parse_array_like() {
		use crate::rich::{Delimiter::*, LiteralKind::*, TokenKind::*};

		tokens![
			URLS,
			(Ident(sym!("urls")), [0, 4]),
			(OpenDelim(Bracket), [5, 6]),
			(
				Literal(Str, sym!("https://paradigm.lighton.ai/api/v1")),
				[8, 44]
			),
			(
				Literal(Str, sym!("https://paradigm-preprod.lighton.ai/api/v1")),
				[46, 90]
			),
			(
				Literal(Str, sym!("https://paradigm-dev.lighton.ai/api/v1")),
				[92, 132]
			),
			(CloseDelim(Bracket), [133, 134])
		];
	}
}
