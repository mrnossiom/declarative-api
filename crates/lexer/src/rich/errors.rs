// TODO
#![allow(unused_variables)]

use macros::IntoDiagnostic;
use session::Span;

#[derive(Debug, IntoDiagnostic)]
#[message("we found an invalid identifier")]
pub(crate) struct InvalidIdent {
	#[label("invalid identifier")]
	pub span: Span,
}

#[derive(Debug, IntoDiagnostic)]
#[message("OVLO (Unidentified lexeme object)")]
pub(crate) struct Unknown {
	#[label("ðŸ›¸")]
	pub span: Span,
}
