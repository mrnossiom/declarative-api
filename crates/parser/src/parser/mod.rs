use crate::{
	error::{UnexpectedToken, UnexpectedTokenInsteadOfKeyword},
	PResult,
};
use dapic_lexer::rich::{Delimiter, Enricher, Token, TokenKind};
use dapic_session::{DiagnosticsHandler, Ident, ParseSession, SourceFile, Symbol};
use std::mem;
use thin_vec::ThinVec;
use tracing::instrument;

mod attr;
mod expr;
mod factory;
mod item;
mod ty;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Spacing {
	/// The token is not immediately followed by an operator token (as
	/// determined by `Token::is_op`). E.g. a `+` token is `Alone` in `+ =`,
	/// `+/*foo*/=`, `+ident`, and `+()`.
	Alone,

	/// The token is immediately followed by an operator token. E.g. a `+`
	/// token is `Joint` in `+=` and `++`.
	Joint,
}

pub struct Parser<'a> {
	pub session: &'a ParseSession,
	/// The current token.
	pub token: Token,
	/// The spacing for the current token
	pub token_spacing: Spacing,
	/// The previous token.
	pub prev_token: Token,

	expected_tokens: Vec<TokenKind>,
	cursor: Enricher<'a>,
}

impl<'a> Parser<'a> {
	#[must_use]
	pub fn from_source(session: &'a ParseSession, source: &'a SourceFile) -> Self {
		let tokens = Enricher::from_source(session, source);
		Self::from_tokens(session, tokens)
	}

	#[must_use]
	pub fn from_tokens(session: &'a ParseSession, cursor: Enricher<'a>) -> Self {
		let mut parser = Self {
			session,
			token: Token::DUMMY,
			token_spacing: Spacing::Alone,
			prev_token: Token::DUMMY,
			expected_tokens: Vec::default(),
			cursor,
		};

		parser.bump();

		parser
	}

	#[must_use]
	const fn diag(&self) -> &DiagnosticsHandler {
		&self.session.diagnostic
	}

	/// Expects and consumes the token `t`. Signals an error if the next token is not `t`.
	#[track_caller]
	#[instrument(level = "TRACE", skip(self))]
	fn expect(&mut self, tok: &TokenKind) -> PResult</* recovered */ bool> {
		if self.expected_tokens.is_empty() {
			if &self.token.kind == tok {
				self.bump();
				Ok(false)
			} else {
				// todo!("recover from unexpected token {}", self.token)
				Err(UnexpectedToken {
					parsed: self.token.clone(),
					expected: *tok,
				}
				.into())
			}
		} else if &self.token.kind == tok {
			self.bump();
			Ok(false)
		} else {
			// todo!("recover from unexpected token {}", self.token)

			// TODO: pass `expected_tokens`

			Err(UnexpectedToken {
				parsed: self.token.clone(),
				expected: *tok,
			}
			.into())
		}
	}

	#[track_caller]
	fn expect_braced<T>(&mut self, mut p: impl FnMut(&mut Self) -> PResult<T>) -> PResult<T> {
		self.expect(&TokenKind::OpenDelim(Delimiter::Brace))?;
		let parsed = p(self)?;
		self.expect(&TokenKind::CloseDelim(Delimiter::Brace))?;
		Ok(parsed)
	}

	/// If the given word is not a keyword, signals an error.
	/// If the next token is not the given word, signals an error.
	/// Otherwise, eats it.
	#[instrument(level = "TRACE", skip(self))]
	fn expect_keyword(&mut self, kw: Symbol) -> PResult<()> {
		if self.eat_keyword(kw) {
			Ok(())
		} else {
			Err(UnexpectedTokenInsteadOfKeyword {
				parsed: self.token.clone(),
				expected: kw,
			}
			.into())
		}
	}

	#[instrument(level = "TRACE", skip(self))]
	fn bump(&mut self) {
		use Spacing::*;
		let (next_token, has_space_before) = self.cursor.next_token();

		self.prev_token = mem::replace(&mut self.token, next_token);
		self.token_spacing = if has_space_before { Alone } else { Joint };

		self.expected_tokens.clear();
	}

	#[instrument(level = "TRACE", skip(self))]
	fn check(&mut self, tok: &TokenKind) -> bool {
		let is_present = &self.token.kind == tok;

		if !is_present {
			self.expected_tokens.push(*tok);
		}

		is_present
	}

	/// Consumes a token 'tok' if it exists. Returns whether the given token was present.
	#[instrument(level = "TRACE", skip(self))]
	fn eat(&mut self, tok: &TokenKind) -> bool {
		let is_present = self.check(tok);
		if is_present {
			self.bump();
		}
		is_present
	}

	/// If the next token is the given keyword, returns `true` without eating it.
	/// An expectation is also added for diagnostics purposes.
	#[instrument(level = "TRACE", skip(self))]
	fn check_keyword(&mut self, kw: Symbol) -> bool {
		self.expected_tokens.push(TokenKind::Ident(kw));
		self.token.is_keyword(kw)
	}

	/// If the next token is the given keyword, eats it and returns `true`.
	/// Otherwise, returns `false`. An expectation is also added for diagnostics purposes.
	#[instrument(level = "TRACE", skip(self))]
	fn eat_keyword(&mut self, kw: Symbol) -> bool {
		if self.check_keyword(kw) {
			self.bump();
			true
		} else {
			false
		}
	}

	/// If the next token is the given keyword, returns `true` without eating it.
	/// An expectation is also added for diagnostics purposes.
	#[instrument(level = "TRACE", skip(self))]
	fn check_ident(&mut self) -> bool {
		// self.expected_tokens.push(TokenKind::Ident(kw));
		self.token.ident().is_some()
	}

	#[instrument(level = "TRACE", skip(self))]
	fn eat_ident(&mut self) -> Option<Ident> {
		self.token.ident().map(|ident| {
			self.bump();
			ident
		})
	}

	#[instrument(level = "TRACE", skip(self))]
	fn parse_ident(&mut self) -> PResult<Ident> {
		let Some(ident) = self.token.ident() else {
			todo!("recover")
		};

		self.bump();
		Ok(ident)
	}

	#[instrument(level = "TRACE", skip(self))]
	fn parse_delimited(&mut self) -> PResult<(Delimiter, ThinVec<Token>)> {
		let mut tokens = ThinVec::default();

		let delim_kind = match self.token.kind {
			TokenKind::OpenDelim(delim) => {
				self.bump();
				delim
			}
			_ => todo!("recover"),
		};

		let mut nesting = 0;

		loop {
			match self.token.kind {
				TokenKind::OpenDelim(delim) if delim == delim_kind => {
					nesting += 1;
				}

				TokenKind::CloseDelim(delim) if delim == delim_kind => {
					if nesting == 0 {
						break Ok((delim_kind, tokens));
					}

					nesting -= 1;
				}
				_ => {}
			}

			tokens.push(self.token.clone());
			self.bump();
		}
	}
}

#[cfg(test)]
mod tests {

	#[macro_export]
	macro_rules! parser {
		($name:ident; $src:literal) => {
			let src = $src;
			$crate::parser!($name; src)
		};
		($name:ident; $src:ident) => {
			let session = dapic_session::ParseSession::default();
			let sf = session.source_map.load_anon($src.into());
			let mut $name = $crate::Parser::from_source(&session, &sf);
		};
	}

	#[macro_export]
	macro_rules! assert_tokenize {
		($method:ident, $src:literal) => {
			#[test]
			fn $method() -> Result<(), dapic_session::Diagnostic> {
				$crate::parser!(p; $src);
				insta::assert_debug_snapshot!(p.$method()?);
				Ok(())
			}
		};
		($method:ident, $variant:literal, $src:literal) => {
			paste::paste! {
				#[test]
				fn [<$method _ $variant>]() -> Result<(), dapic_session::Diagnostic> {
					$crate::parser!(p; $src);
					insta::assert_debug_snapshot!(p.$method()?);
					Ok(())
				}
			}
		};
	}

	assert_tokenize!(parse_delimited, "(bar, baz)");
}
