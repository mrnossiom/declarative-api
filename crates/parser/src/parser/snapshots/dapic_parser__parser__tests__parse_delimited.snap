---
source: crates/parser/src/parser/mod.rs
expression: p.parse_delimited()?
---
(
    Parenthesis,
    [
        Token {
            kind: Ident(
                Symbol(29, "bar"),
            ),
            span: Span {
                start: BytePos(
                    1,
                ),
                end: BytePos(
                    4,
                ),
            },
        },
        Token {
            kind: Comma,
            span: Span {
                start: BytePos(
                    4,
                ),
                end: BytePos(
                    5,
                ),
            },
        },
        Token {
            kind: Ident(
                Symbol(30, "baz"),
            ),
            span: Span {
                start: BytePos(
                    6,
                ),
                end: BytePos(
                    9,
                ),
            },
        },
    ],
)
